### ISSUE TITLE
bug / feature request: macro support: better matchers and timing hints

### ISSUE THREAD

**Issue by inchl** [theNestruo/z80-asm-meter-vscode#134]
Some of the z80-asm-meter.macros do not work as expected. Also I really would like an regex option that allows me to define my own complex line-definition and its ts. This allows me to define an other ts based upon specific comments in the line itself.

        // this does not work!
        {
            "name": "test",
            "instructions": ["; [ts=33"]
        },

        // this would be nice!
        {
            "name": "test",
            "ts": 33
        },

        // this would be great (need that!)
        {
            "regex": "someregex",
            "ts": 33
        }

**Comment by theNestruo** [comment]
Hi! Sorry for the late response.

## 1.

```jsonc
    // this does not work!
    {
        "name": "test",
        "instructions": ["; [ts=33"]
    },
```
That does not work, indeed. `instructions` only allow Z80 instructions. Not timing hints, nor SjASMPlus fake instructions, nor anything else; just Z80 instructions.

However, if you want to specify timings for the macro (instead of the instructions)...

## 2.

```jsonc
    // this would be nice!
    {
        "name": "test",
        "ts": 33
    },
```
...this does work, but the allowed keys for timings are `z80`, `msx` and `cpc`.

## 3.

```
    // this would be great (need that!)
    {
        "regex": "someregex",
        "ts": 33
    }
```
This _may_ work, depending on your macro names. Only the _mnemonic_ of the line is matched against the macro name, so if you have one macro called `"test"`, it will be applied if the line contains: `test`, or `test a`, or `test 23*4,bc` or `test (12, 34)`.

---

Having said that:

1. I have planned the macro defintions (by instructions) to be more permissive: allowing directives, fake instructions, repeat counts, etc.
2. Definitely, the macro definitions (by timing values) should allow the same keys as the timing hints in the comments. Currently, they are missing `m1` (as an alias for `msx`) and `t` and `ts`.
3. Finding macros by RegExp is not as straightforward as it may seem, but it would be definitely an improvement.

I'm working on the next major version of the plugin; I'll probably tackle this three points in that version... but it wil take me some time to release that version.

I think point 2 can be walkaround by using the key `"msx"` instead of `"ts"`; I hope that's enough until the new version.

**Comment by inchl** [comment]
Good to hear!

We are planning to release a demo to the msx community of Kings Valley 
with the sound upgraded.
That would not have been possible without your tool!

Having that said... the regex option is a major feature that we welcome 
very much!

basically to distinguish these two cases:

CASE#1:
    djnz label

CASE#2:
    djnz label    // sample: 14

In case#1 the TS must be 9/14, in case#2 the TS must be 20/25. You might 
think why, well... case#2 is (after parsing) generated as:
    djnz _work
    jp _skip
_work:
    ...sampleLogic...
    jp label
_skip:

So based upon the text "// sample:" in the code line your tool should 
generate an other ts. I hoped to solve this my means of a regex to 
detect this. Maybe a regex is not the correct solution for this problem? 
Also regex might match on multiple code lines, so which one you choose? 
The first one? The last one? The one with the best score (not sure 
how?)....

Looking forward to your next release.

Stephan




Néstor Sancho schreef op 2023-10-04 18:33:

**Comment by theNestruo** [comment]
Hi!

It is a really REALLY specific use case!
Right now, you can use timings hints (`// [ts=11]`), but there is no current way to parse the comment as a macro and then consider it a timing hint...

I think I can do something ad-hoc once I have finished the next major version of the plugin (I'm untangling the mess caused by extension growth, so it will be easier to add additional timing hint parsers, including a regexp based one).

**Comment by inchl** [comment]
Hi,

You are right... its too much. For now I added a 'sdjnz' macro 
instruction to distinguish the normal djnz from the one with the special 
comment-info.
I did not want to write custom instructions but for now its ok.

Néstor Sancho schreef op 2023-10-10 20:07:
> Hi!
> 
> It is a really REALLY specific use case!
> Right now, you can use timings hints (// [ts=11]), but there is no
> current way to parse the comment as a macro and then consider it a
> timing hint...
> 
> I think I can do something ad-hoc once I have finished the next major
> version of the plugin (I'm untangling the mess caused by extension
> growth, so it will be easier to add additional timing hint parsers,
> including a regexp based one).

That regex I only suggested to fix my problem, don't think anyone will 
use it.... :-)

Stephan

**Comment by theNestruo** [comment]
I've changed the title of the issue to be more clear:

- Better matchers in macro support: Currently: only exact mnemonic match. Expected: RegExp, entire line, line comment, etc.
- Greater instruction support in macro definitions: Currently: only Z80 instructions. Expected: anything that can be present in the source code: SjASMPlus fake instructions, Glass negative condition instructions, timing hints, etc.

**Comment by theNestruo** [comment]
Both improvements implemented in 5.0.0, available as pre-release/preview here:
https://github.com/theNestruo/z80-asm-meter-vscode/releases/tag/5.0.0-RC1

Please check [v5.x documentation](https://github.com/theNestruo/z80-asm-meter-vscode/blob/version/5.x/README.md) to find related settings.

**Comment by theNestruo** [comment]
Released to marketplace (as I did a mistake and already published the 5.0.0 wihtout the _pre-release_ check).

Use this setting to add RegExps that will be matched against line comments: `z80-asm-meter.timing.hints.regexps`
